### DISCUSSION TITLE
How to update umijs react 16.4.0 version to React 18.2.0

### DISCUSSION THREAD

**Discussion by Khadhiri786** [umijs/umi#10454] [General]
I have updated my umijs React 16.4.0 version to React 18.2.0 version. But I'm getting error.
`Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: `https://reactjs.org/link/switch-to-createroot``.
Where to add createRoot in project.

**Comment by fz6m** [comment]
Umi v3 does not support `react 18`, you can use `react 17` , Umi v4 supports `react 18` by default

**Reply by Khadhiri786** [reply]
@fz6m  I'm getting issue in config after updating umi version 3.5.23 to 4.0.52
Please refer below screenshot
![image](https://user-images.githubusercontent.com/60467237/220352213-11ed4fe3-579d-464b-8c14-1554e41d11c5.png)

**Reply by fz6m** [reply]
Check your `config/config.ts` , if you using `winPath`, import from the `umi/plugin-utils` :

```ts
import { winPath } from 'umi/plugin-utils' // or `@umijs/max/plugin-utils`
```

**Comment by hasakilol** [comment]
the upgrade is really disgusting for umi beginners
